test/test_distribution.py

```diff
@@ -1215,3 +1225,48 @@ def test_categorical_event_dim_conversion(batch_shape, event_shape):
         funsor.to_data(data, name_to_dim=name_to_dim))
     assert actual_log_prob.shape == expected_log_prob.shape
     assert_close(actual_log_prob, expected_log_prob)
+
+
+@xfail_if_not_implemented()
+@pytest.mark.parametrize("shape", [(), (4,), (3, 2)], ids=str)
+def test_power_transform(shape):
+    transform = backend_dist.transforms.PowerTransform(1.5)
+    base_dist = backend_dist.Exponential(1)
+    d = backend_dist.TransformedDistribution(base_dist, transform)
+
+    data = ops.exp(randn(shape))
+    expected_log_prob = d.log_prob(data)
+
+    name_to_dim = dict(i=-3, j=-2, k=-1)
+    dim_to_name = {v: k for k, v in name_to_dim.items()}
+    x = to_funsor(data, Real, dim_to_name)
+    f = to_funsor(d, output=Real)
+    log_prob = f(x)
+    actual_log_prob = funsor.to_data(log_prob, name_to_dim)
+    assert_close(actual_log_prob, expected_log_prob)
+
+
+@xfail_if_not_implemented()
+@pytest.mark.parametrize("shape", [(10,), (4, 3)], ids=str)
+@pytest.mark.parametrize("to_event", [
+    True,
+    xfail_param(False, reason="bug in to_funsor(TransformedDistribution)"),
+])
+def test_haar_transform(shape, to_event):
```

Comment:
Ditto: try adding a `HaarTransform` test case to `test/test_distribution_generic.py` to test sampling, `to_funsor`/`to_data` conversion and other distribution methods?

Reply:
Can you suggest a fix for this broken `test_generic_distribution_to_funsor()`:

<details>

```
$ FUNSOR_BACKEND=torch pytest -vx test/test_distribution_generic.py -k Haar --pdb
===================================================== test session starts ======================================================
platform darwin -- Python 3.7.0, pytest-6.1.2, py-1.9.0, pluggy-0.13.1 -- /Users/fobermey/opt/miniconda3/envs/pyro/bin/python
cachedir: .pytest_cache
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/fobermey/github/pyro-ppl/funsor, configfile: setup.cfg
plugins: forked-1.2.0, nbval-0.9.6, xdist-2.1.0, benchmark-3.2.3
collected 2244 items / 2211 deselected / 33 selected

test/test_distribution_generic.py::test_generic_distribution_to_funsor[dist.TransformedDistribution( dist.Normal(loc=case.loc, scale=1.).to_event(1), dist.transforms.HaarTransform(dim=-1)) (('loc', 'rand(() + (3,))'),)] FAILED [  3%]
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

case = <test.test_distribution_generic.DistTestCase object at 0x7fb9b08c8cc0>

    @pytest.mark.parametrize("case", TEST_CASES, ids=str)
    def test_generic_distribution_to_funsor(case):

        HIGHER_ORDER_DISTS = [
            backend_dist.Independent,
            backend_dist.TransformedDistribution,
        ] + ([backend_dist.torch_distribution.ExpandedDistribution] if get_backend() == "torch"
             else [backend_dist.ExpandedDistribution])

        raw_dist = case.get_dist()
        expected_value_domain = case.expected_value_domain

        dim_to_name, name_to_dim = _default_dim_to_name(raw_dist.batch_shape)
        with interpretation(normalize_with_subs):
            funsor_dist = to_funsor(raw_dist, output=funsor.Real, dim_to_name=dim_to_name)
        assert funsor_dist.inputs["value"] == expected_value_domain

        while isinstance(funsor_dist, funsor.cnf.Contraction):
            funsor_dist = [term for term in funsor_dist.terms
>                          if isinstance(term, (funsor.distribution.Distribution, funsor.terms.Independent))][0]
E           IndexError: list index out of range

test/test_distribution_generic.py:590: IndexError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Users/fobermey/github/pyro-ppl/funsor/test/test_distribution_generic.py(590)test_generic_distribution_to_funsor()
-> if isinstance(term, (funsor.distribution.Distribution, funsor.terms.Independent))][0]
(Pdb) print(funsor_dist.pretty())
Contraction(ops.nullop, ops.add,
 frozenset(),
 (Unary(ops.neg,
   Binary(ops.log_abs_det_jacobian,
   │Unary(ops._InverseTransform,
   │ Variable('value', Reals[3])),
   │Variable('value', Reals[3]))),
  Contraction(ops.add, ops.nullop,
   frozenset({Variable('_pyro_event_dim_-1_...
   (Normal(
   │ Tensor(
   │  torch.tensor([0.4366315007209778, 0.8093...
   │  (('_pyro_event_dim_-1__BOUND_1',
   │   │Bint[3, ],),),
   │  'real'),
   │ Tensor(
   │  torch.tensor([1.0, 1.0, 1.0], dtype=torc...
   │  (('_pyro_event_dim_-1__BOUND_1',
   │   │Bint[3, ],),),
   │  'real'),
   │ Binary(ops.GetitemOp(0),
   │  Unary(ops._InverseTransform,
   │   Variable('value', Reals[3])),
   │  Variable('_pyro_event_dim_-1__BOUND_1', ...
```

</details>

Reply:
I think the underlying reason is that the way we use a lazy `funsor.terms.Independent` to convert `IndependentDistribution`s to funsors  (via the pattern `funsor.distribution.indepdist_to_funsor`) is sort of fiddly and affects pattern matching, especially when combined with transforms. Now that #402 is done we should try to represent them directly by `ops.expand`ing their parameters to the appropriate `event_shape`.  I'll see if I can get this working in a separate PR.